lib/Models/Cesium.ts

```diff
@@ -608,24 +617,65 @@ export default class Cesium extends GlobeOrMap {
         }
       }
 
+      const allCesium3DTilesets = this._allMapItems.filter(isCesium3DTileset);
       const allPrimitives = this._allMapItems.filter(isPrimitive);
       const prevPrimitives = prevMapItems.filter(isPrimitive);
       const primitives = this.scene.primitives;
 
       // Remove deleted primitives
-      prevPrimitives.forEach(primitive => {
+      prevPrimitives.forEach((primitive, i) => {
         if (!allPrimitives.includes(primitive)) {
           primitives.remove(primitive);
+          if (
+            isCesium3DTileset(primitive) &&
+            allCesium3DTilesets.indexOf(primitive) === -1
+          ) {
+            try {
+              // Remove all event listeners from any Cesium3DTilesets by running stored remover functions
+              const fnArray = this._3dTilesetEventListeners[i].removeFns;
+              for (let j = 0; j < fnArray.length; j++) {
+                fnArray[j](); // run the function
+              }
+              this._3dTilesetEventListeners.splice(i, 1); // Remove the item for this tileset from our eventListener reference storage array
+              this._updateTilesLoadingIndeterminate(false); // reset progress bar loading state to false. Any new tile loading event will restart it to account for multiple currently loading 3DTilesets.
+            } catch (e) {}
+          }
         }
       });
 
       // Add new primitives
       allPrimitives.forEach(primitive => {
         if (!primitives.contains(primitive)) {
           primitives.add(primitive);
+
+          if (isCesium3DTileset(primitive)) {
+            const startingListener = this._eventHelper.add(
+              primitive.tileLoad,
+              () => {
+                this._updateTilesLoadingIndeterminate(true);
+              }
+            );
+
+            //Add event listener for when tiles finished loading for current view. Infrequent.
+            const finishedListener = this._eventHelper.add(
+              primitive.allTilesLoaded,
+              () => {
+                this._updateTilesLoadingIndeterminate(false);
+              }
+            );
+
+            // Push new item to eventListener reference storage
+            try {
+              this._3dTilesetEventListeners.push({
```

Comment:
It's not guaranteed that new 3d tilesets will be at the end of `allPrimitives`. I think to make this work you could replace `this._3dTilesetEventListeners` with a [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map), like `Map<Cesium3DTileset, EventListenerRemover>`. It's also make removing easier.

--------

#### Old ideas
You could change `allPrimitives.forEach` above to `this._3dTilesetEventListeners = allPrimitives.map` and return a new `EventListenerRemover` for a newly added dataset, and for an existing 3d tileset return the old `EventListenerRemover` from `this._3dTilesetEventListeners` (get the index from `prevPrimitives.indexOf(primitive)`. Recreating the array every time also deals with dropping old `EventListenerRemover` (and the above `prevPrimitives.forEach` would have already removed the event listeners).


Another option is to store the event handler remove functions on the 3d tilesets:
```ts
(primitive as any)._terriaProgressEventsRemoveFunctions = [start,finished];
```
But this is not as good as the first solution using Map.

Reply:
Changes from @steve9164 have been incorporated. thanks for the useful review!

I used 
```
private _3dTilesetEventListeners = new Map<
    Cesium3DTileset,
    Event.RemoveCallback[]
  >(); // eventListener reference storage
```

No longer using custom Interface EventListenerRemover.
Event listener removers are accessed using the primitive as the key, instead of iterating.

Fixed the import of `Event`.

Some further simplifications of code for readability.

`3dTilesetEventListeners` is not recreated every time I dont think... would this be better? You mentioned in your change suggestions as an old idea.

Please note some errors unrelated to this PR, started a separate issue at https://github.com/TerriaJS/terriajs/issues/6321